### PR TITLE
Merge from orginal repo

### PR DESCRIPTION
**Pull request by LouisSzeto**

### CODE REVIEW

File: CryptoCoarseFundamentalUniverse.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+using Python.Runtime;
+using QuantConnect.Securities;
+using QuantConnect.DataSource;
+
+namespace QuantConnect.Data.UniverseSelection
+{
+    /// <summary>
+    /// Defines a universe that reads crypto coarse data
+    /// </summary>
+    public class CryptoCoarseFundamentalUniverse : Universe
```
Comment (Martin-Molinero) [inline CryptoCoarseFundamentalUniverse.cs:28]: Cool!

File: CryptoCoarseFundamentalUniverseSelectionAlgorithm.cs [hunk 2]
```diff
@@ -35,23 +37,32 @@ public override void Initialize()
...
+        }
+
+        private IEnumerable<Symbol> UniverseSelectionFilter(IEnumerable<CryptoCoarseFundamental> data)
+        {
+            var universe = new List<Symbol>();
+
+            foreach (var datum in data)
             {
-                foreach (var datum in data)
+                Log($"{datum.Symbol},{datum.Price},{datum.Volume},{datum.DollarVolume},{datum.VolumeInBaseCurrency},{datum.Open},{datum.High},{datum.Low},{datum.Close}");
```
Comment (Martin-Molinero) [inline CryptoCoarseFundamentalUniverseSelectionAlgorithm.cs]: Believe ToString() is already implemented, so can clean up this log? Same for the other algos 👍

File: output/data/crypto/binance/fundamental/coarse/20141203.csv [hunk 3]
```diff
@@ -0,0 +1,2 @@
+BTCUSDC 18N,BTCUSDC,378,0.54660842,206.61798276,377.9,378,377.01,
+ETHUSDT 18N,ETHUSDT,378,0.54660842,206.61798276,377.9,378,377.01,
```
Comment (Martin-Molinero) [inline output/data/crypto/binance/fundamental/coarse/20141203.csv]: - There is a `,` at the end that can be removed? 
- Sample data seems a bit excessive, a couple of days should be enough?

File: DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+                                {
+                                    continue;
+                                }
+
+                                var date = $"{dateTime:yyyyMMdd}";
+                                var high = decimal.Parse(csv[2], NumberStyles.Any, CultureInfo.InvariantCulture);
+                                var low = decimal.Parse(csv[3], NumberStyles.Any, CultureInfo.InvariantCulture);
+                                var price = decimal.Parse(csv[4], NumberStyles.Any, CultureInfo.InvariantCulture);
+                                var volume = decimal.Parse(csv[5], NumberStyles.Any, CultureInfo.InvariantCulture);
+                                var dollarVolume = price * volume;
```
Comment (Martin-Molinero) [inline DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs]: Could replace by using existing `LeanDataReader` I think, avoid all the zip and parsing entirely

File: DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+                                    var newVolume = oldVolume + volume;
+                                    var newDollarVolume = oldDollarVolume + newVolume * price;
+                                    var newHigh = Math.Max(oldHigh, high);
+                                    var newLow = Math.Min(oldLow, low);
+
+                                    dataByDate[date][name] = $"{price},{newVolume},{newDollarVolume},{temp[3]},{newHigh},{newLow},";
+                                }
+                                else
+                                {
+                                    dataByDate[date][name] = $"{price},{volume},{dollarVolume},{csv[1]},{high},{low},";
```
Comment (Martin-Molinero) [inline DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs]: We can remove the last `,` been written?

File: DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+            return true;
+        }
+
+        /// <summary>
+        /// Helper method to read lines in a online file.
+        /// </summary>
+        /// <param name="link">The link of online file to be read</param>
+        public IEnumerable<string> ReadOnlineDocumentLines(string link)
+        {
+            WebClient client = new WebClient();
```
Comment (Martin-Molinero) [inline DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs]: `WebClient` shouldn't be use anymore, there's an extension method in Lean which provides similar functionality

File: DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+                            else if (coarseByDate.ContainsKey($"{baseCurrency[dataTicker]}USDC"))
+                            {
+                                rateToUSD = coarseByDate.ContainsKey("USDCUSD") ? 
+                                    decimal.Parse(coarseByDate["USDCUSD"].Split(",").First(), NumberStyles.Any, CultureInfo.InvariantCulture) :
+                                    decimal.Parse(coarseByDate["USDCBUSD"].Split(",").First(), NumberStyles.Any, CultureInfo.InvariantCulture);
+                                refCurrency = $"{baseCurrency[dataTicker]}USDC";
+                            }
+                            else if (coarseByDate.ContainsKey($"{baseCurrency[dataTicker]}BTC"))
+                            {
+                                rateToUSD = coarseByDate.ContainsKey("BTCUSD") ?
```
Comment (Martin-Molinero) [inline DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs]: Interesting, think this section of code should be it's own method with tests 👍, like `GetDollarVolume()`

`BTCUSD` hard coded here is a bit fishy, why is it required, what about, BTCUSDT, etc?

File: DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+
+                var symbolProperties = ReadOnlineDocumentLines("https://raw.githubusercontent.com/QuantConnect/Lean/master/Data/symbol-properties/symbol-properties-database.csv");
+                foreach (var line in symbolProperties)
+                {
+                    var lineItem = line.Split(",");
+                    if (lineItem.First() == _market)
+                    {
+                        baseCurrency[lineItem[1]] = lineItem[4];
+                    }
+                }
```
Comment (Martin-Molinero) [inline DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs]: Lean has a `CurrencyPairUtil` that could be used directly instead?

File: DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+                            var lines = ReadLines(streamFile);
+
+                            foreach (var line in lines)
+                            {
+                                var csv = line.Split(",");
+                                var dateTime = Parse.DateTimeExact(csv[0].Split(" ").First(), "yyyyMMdd");
+                                if (dateTime < _fromDate || dateTime > _toDate)
+                                {
+                                    continue;
+                                }
```
Comment (Martin-Molinero) [inline DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs]: Should check the date before opening the file I think, will avoid a lot of work which we dont need 👌

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Cool! This is going to be great, leaving a few requests 👍

File: output/data/crypto/binance/fundamental/coarse/20141203.csv (hunk 3 above)
Comment (AlexCatarino) [reply output/data/crypto/binance/fundamental/coarse/20141203.csv]: I think we could reduce the sample data to all symbols of one exchange, e.g. Binance, for just 2-3 days and the daily data of individual securities should not be complete (we are selling this data). 
On the other hand, we should run all symbols for all exchanges to pick up any bug and verify that our calculations of volume in quote currency are correct.

File: DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs (hunk 5 above)
Comment (LouisSzeto) [reply DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs]: Hi Martin! That was left there intentionally, check line 216, a new attribute will be added at the end

File: output/data/crypto/binance/fundamental/coarse/20141203.csv (hunk 3 above)
Comment (LouisSzeto) [reply output/data/crypto/binance/fundamental/coarse/20141203.csv]: Hi Martin! the `,` at the end is actually an attribute with null value.

File: DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs (hunk 7 above)
Comment (LouisSzeto) [reply DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs]: `BTCUSD` is here because we think it might have the best chance for currency exchange rate available, if no &lt;token&gt;-USD/BUSD/USDT/USDC exchange rate is available.

File: DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs (hunk 9 above)
Comment (LouisSzeto) [reply DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs]: I'm not sure how... the crypto data are filed by Symbol, just the first column sorted by date

File: DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+
+            return new Security(
+                SecurityExchangeHours.AlwaysOpen(timezone),
+                config,
+                new Cash(quoteCurrency, 0, 1),
+                SymbolProperties.GetDefault(quoteCurrency),
+                ErrorCurrencyConverter.Instance,
+                RegisteredSecurityDataTypesProvider.Null,
+                new SecurityCache()
+            );
```
Comment (Martin-Molinero) [inline DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs]:
```suggestion
            return new Security(symbol,
                SecurityExchangeHours.AlwaysOpen(timezone),
                new Cash(quoteCurrency, 0, 0),
                SymbolProperties.GetDefault(quoteCurrency),
                ErrorCurrencyConverter.Instance,
                RegisteredSecurityDataTypesProvider.Null,
                new SecurityCache()
            );
```

File: DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+        /// Volume in USD
+        /// </return>
+        public static decimal? GetUSDVolume(
+            decimal? baseVolume,
+            string market,
+            string quoteCurrency,
+            List<Security> existingSecurities
+        )
+        {
+            var targetCurrency = market == Market.Binance ? "BUSD" : "USD";
```
Comment (Martin-Molinero) [inline DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs]: Why is this required, seems like a bug hiding somewhere? I'd expect to see target currency always be USD?

File: DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+                    }
+                }
+
+                foreach (var date in _dataByDate.Keys.ToList())
+                {
+                    var coarseByDate = _dataByDate[date];
+
+                    foreach (var dataSymbol in coarseByDate.Keys.ToList())
+                    {
+                        _existingSecurities[dataSymbol].SetMarketPrice(new Tick { Value = (decimal)coarseByDate[dataSymbol][^2] });
```
Comment (Martin-Molinero) [inline DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs]: - Should avoid the `ToList` calls here, it's iterating twice over the collection?
- Why use `coarseByDate.Keys` if ur later going to access it for the value multiple time bellow `coarseByDate[dataSymbol]`, suggest just using `foreach (var dataSymbol in coarseByDate)` instead
- Should set the security price for all the assets before trying to get conversion rates, else what if the first uses the last security, it's price wouldn't be set?

File: DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+                        {
+                            var volume = _dataByDate[date][dataSymbol][^1];
+                            usdVol = GetUSDVolume(volume, _market, _quoteCurrency[dataSymbol], _existingSecurities.Values.ToList());
+                        }
+                        catch
+                        {
+                            Log.Trace($"No USD-{dataSymbol.Value} rate conversion available on {date}.");
+                        }
+
+                        coarseByDate[dataSymbol].Add(Extensions.SmartRounding((decimal)usdVol));
```
Comment (Martin-Molinero) [inline DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs]: `(decimal)usdVol` this will throw is it's null and it's not being caught?

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]

File: DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs (hunk 11 above)
Comment (AlexCatarino) [reply DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs]: Binance doesn't support USD as quote currency, and most of their pairs are associated with BUSD which is their stable coin to USD.

Comment (Martin-Molinero) [reply DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs]: > Binance doesn't support USD as quote currency

Ah right! it should though? I think once this fix is merged https://github.com/QuantConnect/Lean/pull/6259/files this shouldn't be required

**Review by Martin-Molinero — APPROVED** [review]
Cool!
